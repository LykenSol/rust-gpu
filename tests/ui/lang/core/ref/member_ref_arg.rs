// build-pass

use spirv_std::spirv;

struct S {
    x: u32,
    y: u32,
}

// NOTE(eddyb) `#[inline(never)]` is for blocking inlining at the e.g. MIR level,
// whereas any Rust-GPU-specific legalization will intentially ignore it.

#[inline(never)]
fn f(x: &u32) -> u32 {
    *x
}

#[inline(never)]
fn g(xy: (&u32, &u32)) -> (u32, u32) {
    (*xy.0, *xy.1)
}

#[inline(never)]
fn h(xyz: (&u32, &u32, &u32)) -> (u32, u32, u32) {
    (*xyz.0, *xyz.1, *xyz.2)
}

#[inline(never)]
fn h_newtyped(xyz: ((&u32, &u32, &u32),)) -> (u32, u32, u32) {
    (*xyz.0.0, *xyz.0.1, *xyz.0.2)
}

#[spirv(fragment)]
pub fn main() {
    let s = S { x: 2, y: 2 };
    f(&s.x);
    g((&s.x, &s.y));
    h((&s.x, &s.y, &s.x));
    h_newtyped(((&s.x, &s.y, &s.x),));
}
