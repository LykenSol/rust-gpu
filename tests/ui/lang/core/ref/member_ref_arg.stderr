warning: `#[inline(never)]` function `member_ref_arg::f` has been inlined
  --> $DIR/member_ref_arg.rs:14:4
   |
14 | fn f(x: &u32) -> u32 {
   |    ^
   |
   = note: inlining was required due to illegal (pointer) argument
   = note: called from `member_ref_arg::main`

warning: `#[inline(never)]` function `member_ref_arg::g` has been inlined
  --> $DIR/member_ref_arg.rs:19:4
   |
19 | fn g(xy: (&u32, &u32)) -> (u32, u32) {
   |    ^
   |
   = note: inlining was required due to illegal (pointer) argument
   = note: called from `member_ref_arg::main`

warning: `#[inline(never)]` function `member_ref_arg::h` has been inlined
  --> $DIR/member_ref_arg.rs:24:4
   |
24 | fn h(xyz: (&u32, &u32, &u32)) -> (u32, u32, u32) {
   |    ^
   |
   = note: inlining was required due to illegal parameter type
   = note: called from `member_ref_arg::main`

warning: `#[inline(never)]` function `member_ref_arg::h_newtyped` has been inlined
  --> $DIR/member_ref_arg.rs:29:4
   |
29 | fn h_newtyped(xyz: ((&u32, &u32, &u32),)) -> (u32, u32, u32) {
   |    ^^^^^^^^^^
   |
   = note: inlining was required due to illegal parameter type
   = note: called from `member_ref_arg::main`

warning: 4 warnings emitted

